### ISSUE TITLE
OAuth2 Authentication - Make scope a parameter for postman tests  

### ISSUE THREAD

**Issue by tsnyder-gs1us** [w3c-ccg/traceability-interop#457]
We noticed in the credentials issue postman collections (https://github.com/w3c-ccg/traceability-interop/blob/main/docs/tutorials/credentials-issue/credentials-issue.postman_collection.json) the Get Access Token Test has a hard coded value for scope (scope= resolve:dids issue:credentials). 

While in the Authentication Tutorial postman test (https://github.com/w3c-ccg/traceability-interop/blob/main/docs/tutorials/authentication/authentication.postman_collection.json) does not include the scope parameter.

Some OAuth2 providers require a specific scope value to be passed to get a token. Would it be possible to update the Traceability postman collections to accept scope as a parameter via the newman command line and remove the hard coded scope value?

**Comment by nissimsan** [comment]
Hi @tsnyder-gs1us , 

We have struggled with the scope request in the past. It's an area which isn't properly standardized, so it's not surprising that this resurfaces. And we might need to account for provider nuances. 

It isn't perfectly clear which of the collections is causing you the problem:

- "_Some OAuth2 providers require a specific scope value to be passed to get a token_" sounds like it is the Authentication collection which should be added a request scopes. 
- "_remove the hard coded scope value_" sounds like it is "Credentials Issue" should have those scopes removed. 

I'm going to assume it is the latter, which seems most likely and closest relates to issues we've seen before. 

I fail to see how parameterization would help, though. If we moved those two scopes to be injected my newman, you would still end up failing the test. It would also be an anti-pattern to have the collection deviate from what is actually being tested.

However, I think it would be perfectly acceptable to remove `scope: "resolve:dids issue:credentials"` from the Credentials Issue collection _without_ adding spooky newman workarounds. This collection is about issuance, not API security - I actually think it's a mistake that we even do this. (API security is covered in great lengths in the Conformance suite). If you can confirm this would solve your problem, this should be a simple PR. 

@brownoxford, wouldn't you agree, we should just remove scope requests from interop tests? 

PS: You might also want to look at your oauth2 provider's configuration options - there might be a way you could tweak it to work.

**Comment by tsnyder-gs1us** [comment]
We have confirmed that our OAuth2 Provider (Azure AD) requires a specific scope value to be passed when getting a OAuth token. This is why we suggested making it an environment variable for the postman collection that would be passed to the collection via the newman command line when being executed.

Our initial focus is on passing the Credentials Issue collection, which has a Get Access Token Test which needs to call to get the OAuth token for the other tests in the collection. Without the making the environment variable dynamic not sure how the Get Access Token would work in our case.

**Comment by brownoxford** [comment]
@nissimsan couple of things here that I think are worth calling out:
1. We have discussed and standardized on these specific scope names and per-endpoint requirements for trace interop.
2. Removing these scopes from token requests will cause compliant providers to fail tests.

@tsnyder-gs1us can you give more details on which specific scopes are required by Azure AD OAuth2 provider for token requests?

**Comment by tsnyder-gs1us** [comment]
Here is the article we got from Microsoft explaining how the scope is used. https://learn.microsoft.com/en-us/azure/active-directory/develop/v2-oauth2-client-creds-grant-flow#get-a-token. For example MS Graph requires scope to be https://graph.microsoft.com/.default. 

In our case the required value for scope is in this format api://{apiid_from_azure/.default

**Comment by tsnyder-gs1us** [comment]
@brownoxford  @nissimsan Scope is a pretty standard parameter for OpenID/OAuth2 providers. It should be treated the same way as client_id, client_secret, and audience and be a parameter to the Get Access Token test.

What OAuth Providers are others using with the traceability test suite?

**Comment by brownoxford** [comment]
Including the 'scope' parameter in an access token request is optional, [according to RFC 6749](https://www.rfc-editor.org/rfc/rfc6749#section-4.4.2). If scope is omitted from the request, ["the authorization server must either apply a pre-defined value, or fail the request"](https://www.rfc-editor.org/rfc/rfc6749#section-3.3).

It appears that Microsoft takes the latter approach, and requires `scope` to be present in a token request (and requires scope values to follow a specific format). Other providers, such as Auth0, take the former approach and use a default set of scopes corresponding to all scopes granted to the `client_id` being used. Their logic is that if you have the `client_id` and `client_secret`, [there is no reason to restrict the scopes returned](https://community.auth0.com/t/client-credentials-request-ignores-scope-parameter/18496/2) (this behavior can be modified using scripting).

@tsnyder-gs1us, To give you some background on how things got to be the way they currently are: It is reasonable that an implementation may want to restrict one client to only issuing and verifying VCs, but allow others to engage in presentation exchange. In order to do this, differing scopes would need to be required by the various endpoints, and conformance testing would need to ensure endpoint failure if a required scope was not present. This in turn requires specific scopes to be defined and associated with each endpoint in the interop specification.

@OR13, @nissimsan. Even where Azure AD allows custom scopes, there is still some additional non-standard formatting to scope names that is required. It seems like in order to support Azure AD as an OAuth provider, we would need to allow client configuration to include `scope` just like it currently includes `client_id` and `client_secret`.

Should we re-think inclusion of scopes in the specification and just leave it at requiring OAuth2, with `scope` added as an additional, optional, configuration item? Unless I'm missing something, it seems like the decision to require specific scopes per-endpoint for the purposes of restricting what clients are allowed to do can be left up to the implementation and not be something that is required for conformance...

**Comment by nissimsan** [comment]
@brownoxford, 

> Should we re-think inclusion of scopes in the specification and just leave it at requiring OAuth2, with scope added as an additional, optional, configuration item? Unless I'm missing something, it seems like the decision to require specific scopes per-endpoint for the purposes of restricting what clients are allowed to do can be left up to the implementation and not be something that is required for conformance...

No. The subsequent endpoint calls and how they are protected behind scopes is not what we are discussing here. We are only discussing the Get Access Token call (a.k.a. `POST {{TOKEN_ENDPOINT}}`). 

@tsnyder-gs1us's problem is that on the Get Access Token call on the Authentication Tutorial postman does not request a scope, and Azure insists that it does. 

These are not big changes needed: 
1. Update interop collections to always include a `scope` in Get Access Token calls. 
2. Add to the spec that a `scope` must be requested in Get Access Token calls (debatable if even needed).

@tsnyder-gs1us, I'm curious, does Azure accept an empty `scope` like this: 

```
client_id: "foo"
client_secret: "bar"
grant_type: "client_credentials"
scope: ""
```

If yes, then that could be the solution for the Authentication Tutorial collection.

**Comment by tsnyder-gs1us** [comment]
@nissimsan Azure AD requires a specific scope be passed. If no scope is sent it will return an error. So long term to support Azure AD any postman collections that has a Get Access Token will need to have a variable for scope. 

In the short term we are only focus on the Credentials Issue collection. If that collection accepts scope as a parameter we could move forward.

**Comment by brownoxford** [comment]
@nissimsan All endpoints that require a caller to "have" a scope verify this requirement by checking that the presented bearer token includes that scope; in that sense, you are right that the Get Access Token call is the only one that matters here because a bearer token with a scope is ALWAYS generated by a get token request.

That being said, all of the other endpoints under test that require a certain scope be present in the bearer token make a request to obtain a token with the required scopes first. For example, when testing `/credentials/issue` in the conformance suite, a call is first made to the auth token endpoint to get a token with "issue:credentials" scope.

While it is definitely an issue for Azure that no scope is present for the token request in the Authentication tutorial, the larger issue is that Azure requires a certain format for scope names that we do not currently support, and I don't see us moving to that format and telling everyone else that they need to use scopes in Azure format.

The suggestion put forth by @tsnyder-gs1us is that we allow implementations to provide scope as a configuration variable. This makes it much more difficult to test per-endpoint scope requirements (configuration explosion), and leads to the proposal that we omit per-endpoint scope requirements from conformance testing and instead required that each configured client have permission to perform all actions in the system under test. In some cases, e.g., Azure AD, this *requires* that a scope be provided to the auth token request. In others, like Auth0, this is the default behavior if no scopes are presented during the auth token request.

**Comment by brownoxford** [comment]
Note: as @tsnyder-gs1us pointed out, the vc-api test suite does allow per-endpoint specification of scopes, [see the mesur.io configuration for examples](https://github.com/w3c-ccg/vc-api-test-suite-implementations/blob/main/implementations/Mesur.io.json). 

@tsnyder-gs1us to respond to your [previous comment](https://github.com/w3c-ccg/traceability-interop/issues/457#issuecomment-1309034741)

> To clarify the specific scope is only required for getting the OAuth2 token. Different scopes can be used for the issuer/verifier.

In the vc-api test suite, the scopes listed in `issuers` are passed to the token endpoint when getting a token for making requests to `/credentials/issue`, and the scopes listed in `verifiers` are passed to the token endpoint when getting a token for making requests to the `/credentials/verify` endpoint.

Scopes are always passed to endpoints by including them in a bearer token issued by the OAuth provider, they are never passed just passed directly to the endpoint by the requestor. A service protected by OAuth trusts the OAuth provider, not the entity making the request.

**Comment by nissimsan** [comment]
Alright, understood - the problem stems from Azure's `.default` scope names which are for some reason required for client_credentials, not just a naming convention. That does indeed seem difficult to support without pretty significant rework. 

Just to get all options on the table: 

- One is to back away from the narrow/granular scopes which we've agreed on and just have one, which we make configurable. 
- But I suppose we _could_ also make all our scopes configurable. So, eight parameters instead of one. One vender can fill in `issue:credentials` another `credentials/issue/.default`. This way we could keep our current granularity.

**Comment by tsnyder-gs1us** [comment]
For the VC-API Test Suite we verified that the oauth2 configuration supports passing a **scopes** parameter. We have been able to successfully use this to authenticate with Azure AD and run the VC-API Test Suite locally against our VC-API Implementation.  

While it may seem like overhead the more configurable settings is the best long term option. 

"oauth2": {
       "scopes": [api://{apiid_from_azure/.default]
  },

**Comment by OR13** [comment]
@tsnyder-gs1us are you proposing a change? I want to make sure we capture the action here.

**Comment by tsnyder-gs1us** [comment]
> @tsnyder-gs1us are you proposing a change? I want to make sure we capture the action here.

Yes, I am recommending that scope be configurable for the Get Access Token postman tests. Instead of having the scope value hard coded it should be setup as a postman environment variable and passed into the tests from the secrets collections.

For Example: 
 scope: ${{secrets[format('{0}_SCOPE', matrix.actor)]}}

![image](https://user-images.githubusercontent.com/81628873/201954998-634580b7-c805-4f31-b849-b417e1a14eb0.png)

**Comment by OR13** [comment]
@tsnyder-gs1us would you be willing to do a PR changing this in one simple test, for review?

@ipbyrne @brownoxford what do you think about this?

**Comment by brownoxford** [comment]
@OR13 I'm in favor of removing all scope-specific requirements from the specification and falling back to simply requiring OAuth2 authentication - this would leave it up to implementations to decide whether to require specific per-endpoint scopes, and would require that client_credentials grants used in conformance/interop testing include ALL scopes required to interact with all endpoints under test.

In some cases, e.g., Auth0, this is the default behavior. In others, e.g., Azure AD, specific scopes may need to be included in the auth token request. For those cases, the addition of a new configuration item for scope is necessary.

**Comment by OR13** [comment]
@brownoxford your proposal will lead to each implementer defining their own scopes (no interoperability at this layer), and the configuration for each actor having to manage that (additional registration / maintenance complexity).

Isn't it simpler to just allow for an actor to add additional scopes from the env?

Won't any scopes that are not understood automatically be ignored?

**Comment by brownoxford** [comment]
@OR13

> your proposal will lead to each implementer defining their own scopes (no interoperability at this layer),

Agreed. Implementers will need to clearly outline required scopes in their API documentation, and consumers would need to take care to use the necessary scopes when making auth token requests.

> Isn't it simpler to just allow for an actor to add additional scopes from the env?
> Won't any scopes that are not understood automatically be ignored?

I can only speak to the behavior of Auth0, where requests for unknown scopes results in a 403 response as in this response to a request with scopes `resolve:dids foo:bar`
```
{"error":"access_denied","error_description":"Client has not been granted scopes: foo:bar"}
```
The primary issue here is that we cannot currently support Azure AD because it is impossible to manipulate Azure AD into using our desired scope names.

**Comment by brownoxford** [comment]
One thought I had was that we could add normative language stating that OAuth providers MUST grant all scopes assigned to a client ID when a token is requested, which would make the whole scope configuration opaque - the problem with that is Azure AD *requires* a scope value in the token request body.

**Comment by OR13** [comment]
@brownoxford ah yes, you are right regarding scopes in requests for a token... I was thinking about another layer.

I still think its probably best to solve for this without destroying the interoperability at the scope layer, by allowing for an env value which can be "empty" to be included in the token request... for example:

```
scope: "resolve:dids ${AZURE_AD_SCOPE}"
```

When not set, there is no change to the tests, and when set, the token will be returned with proper scopes... right?

The question is does Azure AD know not to error when you request a scope like "resolve:dids api://{apiid_from_azure/.default"... 

... and shouldn't it error if that scope (resolve:dids) is not bound to the client (similar to how Auth0 errors) ?

**Comment by brownoxford** [comment]
To clarify, `resolve:dids` will never be bound to a client in Azure AD as it is an invalid scope name.

Couple of thoughts related to broken interoperability in the case of an Azure AD implementation:

- We cannot include normative text saying that the `${AZURE_AD}` scope is required in auth token requests
- Any client not including the `${AZURE_AD}` scope will not be able to get an auth token
- Any client including an unrecognized scope like `resolve:dids` might get an error like the one that Auth0 returns (depending on how Azure AD works)
- Interoperability is therefore broken for Azure AD implementations

**Comment by OR13** [comment]
This post seems relevant: https://medium.com/@eriksonmurrugarra/azuread-oauth-2-0-client-credentials-with-custom-scope-be2bfb924885

**Comment by brownoxford** [comment]
@OR13 @ipbyrne this is actually the article that I read that demonstrated how this does not work with Azure AD. In this article, note that the scope request needs to be `api://my-test-application/.default`.

**Comment by nissimsan** [comment]
Putting here in writing what I suggested on the call...

We currently have 6 secrets per provider which are passed into the tests as parameters. 
And we have 8 defined scopes. 

If we just add one scope as parameter, then we loose the security granularity which we've gone through so much effort to arrive at and are testing rigorously in the conformance suite. 

But if we add 8 scopes as individual parameters, then we could keep the granularity. We wouldn't test the name of those scopes, but we could still test granular endpoint access. 

We would likely want to keep these 8 parameter values in plain text somehow, not secrets.

**Comment by tsnyder-gs1us** [comment]
> @OR13 @ipbyrne this is actually the article that I read that demonstrated how this does not work with Azure AD. In this article, note that the scope request needs to be `api://my-test-application/.default`.

We confirmed with Microsoft that Azure AD does require /.default be passed to get OAuth Token.

**Comment by nissimsan** [comment]
Looking for Microsoft to weigh in before deciding on path ahead.

**Comment by OR13** [comment]
From our meeting minutes for today's call:

> Ben - Transmute says: Orie: I pinged my friends at Microsoft to weigh in on this issue. I want a response from them before acting on this issue. 

> Ben - Transmute says: Orie: I think we want to say that if you're capable of validating, you still get some coverage. We want to have some security posture around that. And then we add a description to the respec document around how scopes are handled in the profile around interop. 

> Ben - Transmute says: Orie: And then we would need to have some changes in the tests so that people don't have to jump through hoops. So update the respec document, and then add a flag to the tests to ignore scopes for other OAuth implementations. 

TLDR: We are likely to add a warning about Azure AD and scopes to resolve this issue, if we don't get concrete feedback.

**Comment by brownoxford** [comment]
@OR13 @tsnyder-gs1us do we have any further information from Microsoft, or should we proceed with wording on a potential warning to add to our repo?

**Comment by brownoxford** [comment]
Re-review and make concrete proposal to rewrite tests so that we are not testing for specific scopes.

**Comment by ipbyrne** [comment]
Revisiting this. Could we update the request/collections so they use the hard-coded scopes unless something like ${{secrets[format('{0}_SCOPE', matrix.actor)]}} is provided in the postman environment using the collection? This way for people you need to modify the default scopes can without it breaking things for everyone who can use the default scopes?
